### ISSUE TITLE
Terminal gets extremely sluggish after a day of use

### ISSUE THREAD

**Issue by lminer** [microsoft/terminal#7710]
I'm finding that if I use Windows Terminal for more than a day, it starts to get extremely sluggish: there is a lag in opening and closing new tabs and a lag in switching between tabs and a lag in typing. This problem immediately resolves itself if I close the application and restart it.

I'm not sure how to give a more detailed bug report. I tried downloading WPR, but it errored out when I tried to make a recording.

My setup is as follows: I'm using the beta version of Terminal, running WSL2 with ubuntu 20.04. I'm on the insider branch of Windows 10 with CUDA support for WSL2 enabled. My laptop is a Razer Blade advanced 2019 4K with an nvidia 2080 mobile graphics card.

**Comment by mehgcap** [comment]
Just for the record, I'm seeing this as well. I opened a duplicate issue, not realizing this one was already here. I'll try
 to capture a dump next time it happens.

**Comment by alexandermalfait** [comment]
> @lminer (or anyone else seeing this): A dump would be helpful in trying to find the root cause for this issue. You can capture one using Task Manager by right-clicking on WindowsTerminal.exe in the _Details_ tab. Thanks!

Do you mean the "create dump file" option?

**Comment by zadjii-msft** [comment]
![image](https://user-images.githubusercontent.com/18356694/97019292-c5164680-1515-11eb-9781-562150353815.png)

Yep, that one

**Comment by lminer** [comment]
How do I send it to you? The files are big

**Comment by alexandermalfait** [comment]
@zadjii-msft , you can download a dump file and a video showing the issue from this link: https://we.tl/t-Q9SG7e5z6r

Note that the dump file is 1GB uncompressed, which is a lot of memory usage for a terminal.

In task manager, I could see 700MB of memory usage, dropping to 70MB after restarting the terminal.

**Comment by mehgcap** [comment]
@zadjii-msft I can only get this to happen reliably when SSH-ed into a few work servers. In case the dump file contains potentially sensitive data, is there an email address where I can send it?

**Comment by djsavvy** [comment]
@mehgcap see https://github.com/microsoft/terminal/issues/7981 -- @DHowett provided an e-mail where you can send a dump.

**Comment by rebane2001** [comment]
I can confirm it as well, here's a GIF of me typing a sentence in a Windows Terminal vs just WSL
![2020-10-30_22-26-22](https://user-images.githubusercontent.com/6316541/97753761-f8049f80-1afe-11eb-9021-b508e2f0afd5.gif)

Usually, when using the Windows Terminal, I have a SSH session through WSL into a tmux session on a server. This server probably prints over a million lines of text a day, so my gut tells me that this data somehow gets stored in a cache/history of some sort that gets laggy once large enough. Of note is also that tmux sends a lot more data than a simple SSH session due to how it works. However, closing tabs does not solve this issue, I need to restart Windows Terminal for it to run fast again.

**Comment by alexandermalfait** [comment]
I've just had a (long overdue) 4 day holiday. 

Coming back to my desktop session, Windows Terminal was extremely slow.
So the issue seems to "build up" even when WT is not being used at all.

**Comment by stedaniels** [comment]
I get this consistently as well, and have for months.  I know I shouldn't be leaving terminal windows open over night, but it really shouldn't be a problem either :-)  This seems to happen on multiple unrelated Windows 10 installs.  E.g. a work AD joined workstation and also my personal Dell PC, and personal homebuilt 7 year old PC.
Do you need any further dumps sending to you?
I'm pretty sure if I just open Windows terminal on a Friday evening, and attempt to use it on a Monday morning (wishful thinking) then the issue would also be present.
Cheers,
Steve

**Comment by RoyLi-Pvt** [comment]
I have same problem

**Comment by kyuyeonpooh** [comment]
I also have the same problem.

**Comment by NeilMacMullen** [comment]
I consistently see this issue (raised as #8640 above) on a machine we use to run some production scripts.  Will report further findings here.

**Comment by MikeRixWolfe** [comment]
Same issue in both 1.4 release and 1.5 prerelease with WSL1 (Ubuntu 18.04) on my Kaby Lake era desktop build.

**Comment by kyuyeonpooh** [comment]
I found that **no issue occurs when using terminal with only single tab**. I have opened a single-tab terminal for 2 weeks, but no speed problem found while opening new tab or tab switching. **Opening multi-tab terminal for a long time** may be the cause of this issue.

**Comment by Blaok** [comment]
> I found that **no issue occurs when using terminal with only single tab**. I have opened a single-tab terminal for 2 weeks, but no speed problem found while opening new tab or tab switching. **Opening multi-tab terminal for a long time** may be the cause of this issue.

My experience is that keeping a single tab only makes the problem shows up slower. After a couple of days it still becomes sluggish when I try to copy from the terminal. My workaround now is to keep a single terminal tab, launch a tmux server in WSL, switch sessions in tmux, and restart the terminal when it's sluggish.

**Comment by Don-Vito** [comment]
As @yakobyd mentioned here https://github.com/microsoft/terminal/issues/8702#issuecomment-754303204, the problem seems to reproduce consistently with ssh+tmux. His observation that can be relevant was that:

>When printing a large text file during a tmux session, the memory consumption of the terminal process doubles (60MB -> 120MB). After closing the tmux session, and clearing the terminal, there is no way to access the printed text. However, the memory of the terminal process does not clear.

**Comment by stedaniels** [comment]
I didn't mention it at the time, but I use tmux heavily.  It seems to be a common theme?

**Comment by mehgcap** [comment]
Just for the record, I do *not* use TMUX or any other multi-window solution in my terminal work. I sometimes ctrl-z something and use fg to get back to it, but even that is pretty rare. That's as far as I go into having multiple things going at once in the same session. I tend to pop open a new terminal and get into the same server if I need to do multiple things on one host.

**Comment by djsavvy** [comment]
Interestingly, I haven't been able to reproduce this over the last couple of weeks. I've left my terminal session open (connected to ssh/tmux/mosh) for several days, and switching tabs/typing is still snappy. 

I'm on the latest Windows Terminal Preview. 

Is it possible that a Windows update fixed this for me? I'm on the latest insider dev build.

**Comment by JohnLukeBentley** [comment]
Windows terminal is extremly sluggish/slow for me too. As for @stedaniels I've had this behaviour for months. 

Unlike others the sluggishness comes on quickly. From opening Windows Terminal I try to type (each character individually, without using autocomplete) and execute the powershell command `Get-ChildItem | Format-Table -Autosize` repeatedly. By the third iteration it's unbearably slow. At this point WT is unuseable.

I'm on WT Version 1.4.3243.0. Using Powershell 7.1.1 (64 bit), installed from an msi. Sluggishness also occured with Powershell installed via the MS Store (which I presume was also 64 bit).

A random stab in the dark at an interaction causing the problem: Is the 64 bit version of Powershell causing the slowdown? Before I get around to testing this myself perhaps anybody else can say whether their sluggishness in WT exists with a 32 bit Powershell install.

**Comment by RoyLi-Pvt** [comment]
7Tabs 4SSH&TMUX 24Hours no lag on input but tab switch is very slow~

**Comment by JohnLukeBentley** [comment]
I've just uninstalled Powershell 64bit (7.1.1) and Installed Powershell 32bit (7.2.0-Preview 2). On repeat of my previous steps, typing out in full and executing `Get-ChildItem | Format-Table -Autosize` three times, none of the sluggishness of Windows Terminal exists. Ctrl+Tabbing between tabs is smooth and fast.

I then uninstalled Powershell 32bit (7.2.0-Preview 2) and installed Powershell 32bit (7.1.1). I repeated the same tests. The sluggishness was also absent.

This was only an test of sluggishness on immediately opening Windows Terminal. I'll try leaving Windows Terminal open for a day to see if I get any of the sluggishness others report.

So, at this point, it would seem the interacting cause (for immediate sluggishness at least) is the 64bit version of powershell (plus some additional factor or factors). The (temporary) workaround is to use the 32bit version (e.g. from https://github.com/PowerShell/PowerShell/releases > Some release > Assests > PowerShell-7.1.1-win-x86.msi ).

@lminer, and others, does your sluggisness (appearing only after some while) with Windows Terminal exist with a Powershell 32bit version installed (and with Powershell 64bit removed)?

**Comment by rebane2001** [comment]
The 64bit PowerShell issue is probably separate from what others have been experiencing as I have this problem all the time with just WSL and not PowerShell. I also couldn't reproduce the PowerShell sluggishness.

**Comment by JohnLukeBentley** [comment]
@rebane2001 but, even if you aren't opening a Powershell tab in WT, you'll have Windows Powershell Legacy (e.g. 5.1) *installed* (given it is baked into Windows).

When I experienced the slugishness it was in any kind of Windows Terminal command line tab: Powershell Core; Powershell Legacy; and even Windows Command Prompt (cmd.exe), e.g. when typing and executing `ipconfig`. This is now removed for me.

If you install a Windows Powershell core 32 bit version (e.g. https://github.com/PowerShell/PowerShell/releases/download/v7.1.1/PowerShell-7.1.1-win-x86.msi) and execute Linux commands from Windows Terminal from whatever command line you like, e.g. Windows Command Prompt does that remove the sluggisness you've been experiencing?

(This is assuming I haven't misuderstood that with WSL installed and enabled when using Windows Terminal you don't get some Linux specifc command line. But in that case the question would remain: does having a Windows Powershell core 32 bit version installed neverthless remove the sluggishness you are experiencing?)

**Comment by JohnLukeBentley** [comment]
After leaving my Windows Terminal open for more than 24 hours (after an uninstall of Powershell 64 bit and install of Powershell 32 bit) no sluggishness has reappeared. Nor is Tab switching  (just two tabs) slow.

**Comment by stedaniels** [comment]
@Don-Vito do you still need a repro for this, or have you managed it yourself yet?

**Comment by Don-Vito** [comment]
> @Don-Vito do you still need a repro for this, or have you managed it yourself yet?

@stedaniels - no I didn't. A repro would be lovely :blush:

**Comment by JohnLukeBentley** [comment]
Alas the update of Windows Terminal (automatically through the store) to 1.5.10271.0 has made the sluggishness return, with Powershell 32 bit installed.

**Comment by JohnLukeBentley** [comment]
The sluggishness is worse in Windows Terminal (preview) 1.6.10272.0.

**Comment by NeilMacMullen** [comment]
I can't prove it (yet) but I think the newer version may spontaneously shut down after outputting large amounts of text to a tab.

**Comment by wrossmann** [comment]
One thing I've noticed regarding this issue is that the length of time it takes to present itself as a problem seem to be related to how much output is happening in the tabs.

For example, my primary use is to have irssi running on the foreground tab, and then have a couple tabs for the odd command that I need to run. I'd say that irssi does a "medium" amount of output, the usual backgrounded tabs do virtually none, and I've left terminal running for weeks on end like this with no noticeable slowdown. However, recently I've been having some issues with my internet and I've been leaving mtr running in a background tab to keep a rough eye on packet loss. Within a day or two of running mtr with its "high" amount of output terminal starts to lag very noticeably and needs to be restarted.

I know that this is purely anecdotal, but it may be an avenue of investigation. It could be the general volume, background volume, or maybe something that ncurses does that tickles terminal in the wrong way.

Thanks for coming to my TED Talk.

**Comment by rebane2001** [comment]
I can confirm what @NeilMacMullen is saying - I've had the Terminal crash after running for a while. Often when the Terminal is on the brim of crashing, resizing the window forces a crash.

**Comment by NeilMacMullen** [comment]
@rebane2001   That is interesting.  I was beginning to think the "crash" was actually  due to WIndows Store performing a background update (which forces any open terminals to close - see https://github.com/microsoft/terminal/issues/3915).

**Comment by rebane2001** [comment]
I am fairly certain that my crashes have not been caused by updates

**Comment by JohnLukeBentley** [comment]
My sluggishness has disappeared with WT preview version: 1.6.10412.0 (autoupdated through Windows Store).

A point relevant to #3515 (as I've previously mentioned I can't tell if these two issues are two faces of the same cause).

**Comment by desmap** [comment]
On 1.6.10412.0 and I was about to post a similar video as in https://github.com/microsoft/terminal/issues/7710#issuecomment-719803095 from @rebane2001.

I am using Windows Terminal as my daily driver for some weeks now. Today, I had to finish some procrastinated work and somehow , I intuitively just closed Terminal and started WSL2. What can I say, it felt like back home and I was finally working. 

FWIW, then I restarted Windows Terminal to check if this lag is still there. And it is, so it even doesn't has to run for long. It's subtle but enough to annoy you. The bigger the window the more it's obvious. If I add transparency via AHK even more but not at all with WSL2 with AHK-added transparency. Even if you put both Terminal and WSL2 above each other, turn on AHK transparency and do quick cursor movements in vim, you see the lag better, even if you do (send) the movements in Terminal (you need tmux for this btw).

FWIW2, I work on a remote server via ssh and with tmux.

**Comment by Cuberick-Orion** [comment]
Similar issue here. My usual scenario is: multiple tabs on `wsl2`, each with an `ssh` connected to a remote server, with (usually) many outputs.

In addition to the Terminal being sluggish, I also notice a constantly increasing memory usage by its main process. For 4-5 days the usage could grow to 2GB in total.

Would appreciate any comments.

**Comment by mengx8** [comment]
I have same problem 1.6.10571.0

**Comment by bartoszek** [comment]
I'm using `tmux` session inside `WSL:Ubuntu` image and after detaching session `[ctrl-a]d`, restarting `Terminal:16.10571.0` and reconnecting with `tmux a` speed is back at normal and memory usage is back at reasonable `~30MB` 
This issue is really pain full, after few days typing speed goes down to like 2char/sec on my thinkpad w530 😢 ( how is that `cygwin` is able to deliver with all it's overhang and native application isn't :feelsgood: )

**Comment by zadjii-msft** [comment]
You know, if we knew what was causing this one, we would have solved it by now. Unfortunately, we still don't have a good repro, and we don't have any traces of a terminal in this state, so there's not much to work with 😕

**Comment by alexandermalfait** [comment]
@zadjii-msft , I did [send you a dump](https://github.com/microsoft/terminal/issues/7710#issuecomment-716423695) on october 26th

Did that not help to find a root cause for this issue? Seems to me like it's memory leak.

**Comment by zadjii-msft** [comment]
holy crap there's a dump

Sorry I was on paternity leave at the time so I was not regularly checking mail. That's what I get for checking mail on "vacation" 😆 

Would you mind sending it again? I think the link expired?

**Comment by alexandermalfait** [comment]
No problem! We can all agree babies are vastly more interesting than memory leaks :)

Indeed the link has expired, and I'm searching high and low but can't the original file anywhere. 
I restarted my terminal this morning, I'll wait a few days for the issue to pop back up and send you a new dump.

**Comment by alexandermalfait** [comment]
PS: is there a way I can send you the link privately? 
I guess the dump will contain terminal output, which will probably have sensitive data in it.

**Comment by zadjii-msft** [comment]
Sure thing, my email is in my profile ☺️

**Comment by desmap** [comment]
Tried Terminal again after a while, wanted to do some comparison/screen recording, all was good and before moving on I went to the kitchen to get me some coffee.

When I came back, my notebook fans were to the max, so opened my task manager and saw this on the "3D" chart:

![image](https://user-images.githubusercontent.com/43666255/111770950-5910f200-88ab-11eb-8556-9b5b88e559c1.png)

So, I slowly closed first Chrome, then Spotify and eventually, Terminal, take a look: 

![image](https://user-images.githubusercontent.com/43666255/111771338-d5a3d080-88ab-11eb-8e30-4722598c3f44.png)

**Comment by alexandermalfait** [comment]
@zadjii-msft , I've sent you a new dump over WeTransfer, dit you receive it?

Primary "symptom" I'm seeing is slow tab switching, like a second of delay. After restarting Terminal, tab switching is instant again.

**Comment by deepend-tildeclub** [comment]
Sounds like my issue.. Most times I only have one tab open with a SSH session to my server and windows terminal will be using 1.6GB of ram and very slow to respond to clicks and eventually slow to even type in the terminal. 
My pc is not slow its 8cores 16threads with 32GB ram, NVME SSD.. So don't understand why it would get sluggish.

**Comment by DS992** [comment]
Experiencing the same issue using WT Preview V 1.7.572.0. I am also a heavy user of tmux (through WSL2).

Clearing the history does nothing. Killing tmux does nothing. It gets slower over time and/or due to the amount of text output in the terminal. For me at least.

Opening a new window (not tab) of WT Preview seems to make it work normally, but the other Windows (which I kept open) is still "sluggish".

So this is not a critical issue, but it's extremely annoying.

**Comment by BryceEakin** [comment]
Reported the same issue separately - my mistake.  Here's my summary and diagnostic data.  Hope it helps get this resolved.  Definitely an annoyance, though Windows Terminal has been a huge win in my workflow.

Windows Feedback link for diagnostic data: https://aka.ms/AAbj0dj

**Comment by miniksa** [comment]
I think I have an idea on this. I need to capture allocation traces for `DispatcherTimer`s as I have a 2GB dump from this state that shows almost 2 million of those allocated and alive in the heap.

**Comment by miniksa** [comment]
`ThrottledFunc<>` leaks `DispatcherTimer` instances in two spots:

TerminalControl!<lambda_8050e662c890fd8b88061e60f2bb6336>::operator()+0x6d [E:\BA\3\s\src\cascadia\TerminalControl\ThrottledFunc.h @ 62]

and 
TerminalControl!<lambda_922c48371e5965b2d4b2a51a49c8c9c3>::operator()+0x6d [E:\BA\3\s\src\cascadia\TerminalControl\ThrottledFunc.cpp @ 41]

Something about the pattern they're using is making an extra ref/instance that never gets let go and just consumes more and more over time...

**Comment by yamlCase** [comment]
This sounds very promising!  I'm getting about 6 hours of work before I have to `wsl --shutdown` so would be able to tell almost immediately if there's improvement.  Would it be possible to provide a hotfix build?

**Comment by zadjii-msft** [comment]
Well, the next release is already scheduled for sometime next week, so probably not? We'd want to selfhost it internally too and make sure we didn't accidentally make this worse 😅

**Comment by panicfarm** [comment]
This bug is very serious. I Many programmers need the terminal running for a long time in order not to lose state, but it becomes unusable after several days.

**Comment by zadjii-msft** [comment]
@panicfarm then you'll be happy to know that the root cause was found and should be fixed in the next preview build ☺️

**Comment by NeilMacMullen** [comment]
Fantastic ! Great work team-terminal :-)

**Comment by desmap** [comment]
Thanks for the fix and update. Just tried it briefly and compared to wsl2's terminal (btw, how is this terminal actually called?): I opened a remote server, vim in tmux and just quickly pressed `j` and `k` repeatedly and very fast with the line highlighted in vim (`:hi cursorline something...` ). Terminal was fullscreen on a 4k 14". The difference is quite obvious: While this up and down is in wsl2 super responsive and crisp, even on a remote server, even on a 4K res with a crappy embedded GPU it lags on Windows Terminal 😑.

Not sure if this is the right thread and sorry that I don't have better news. Didn't even bother to install the update on my new machine because lagginess is a deal breaker, I mean do you guys remember https://news.ycombinator.com/item?id=21862087 whatever, maybe it's just my machine...

**Comment by mehgcap** [comment]
So far, so good, at least on an SSH connection I left open to a work server for 24 hours. I did notice that things quickly got sluggish in the settings window, of all places, though I have no idea if the cause is the same. Still, at least the actual terminal part seems to be working much better. Thank you to those who put in the work to solve this one.

**Comment by YAMLcase** [comment]
I noticed a marked improvement, but still find my terminal getting into sluggish mode.  I most frequenly notice it when lots of lines displaying on a remote ssh session (i.e., cat big logfile).  the latest WSLTTY doesn't get sluggish with the same test.

**Comment by hinst** [comment]
I still get this problem, Windows Terminal becomes very slow after 24+ hours

**Comment by lhecker** [comment]
@hinst This issue is really old. You should open a new one. If you can repro it so easily, it would be great if you could do some initial investigation for us (since this doesn't happen to us maintainers). If you know how to use System Informer for Process Explorer, the CPU usage of individual threads would be interesting. Additionally, a list of loaded modules would be interesting. Lastly, a WPR recording would be helpful. If you need assistance with any of these, we should discuss that in the new issue.
